Install pre-compiled files into source tree

will-v-pi opened the pull request:
This adds an install step in the external projects when `USE_PRECOMPILED=false` to copy the compiled binaries into the source tree.

This also adds a new action to check that the pre-compiled files have been updated correctly, although I'm not sure if we should add that or not, as the files will change slightly depending on the compiler/SDK used so that action will probably fail quite a lot?

Fixes https://github.com/raspberrypi/picotool/pull/240#discussion_r2129222486

will-v-pi committed:
Copy newly compiled files over when using USE_PRECOMPILED=FALSE

will-v-pi committed:
Add action to check precompiled binaries

will-v-pi committed:
Fix build with precompiled

will-v-pi committed:
Update precompiled binaries

lurch commented on .github/workflows/check_precompiled.yml:
```diff
@@ -0,0 +1,81 @@
+        uses: actions/checkout@v4
+
+      - name: Install dependencies (Linux)
+        if: runner.os == 'Linux'
```
Is `if: runner.os == 'Linux'` needed, given that you also have `runs-on: ubuntu-latest` ?

will-v-pi committed:
Remove `if: runner.os == 'Linux'` where not needed

Keep the step name "Install dependencies (Linux)" in case dependencies change later, to make changing all of them simpler

lurch commented:
Looks like there are merge conflicts (which are probably what's also causing the CI failure?)

will-v-pi commented:
That CI failure is why I'm thinking this change doesn't want to have the check action, as the only change has been in the SDK so we don't really need to update the `.elf`

The conflict is just because the code for those files has been updated

will-v-pi committed:
Change workflow to just check for file updates, rather than identical files

lurch commented on .github/workflows/check_precompiled.yml:
```diff
@@ -23,50 +26,65 @@ jobs:
+            if echo "$updated_files" | grep -q "$file"; then
+              echo "File $file is in the PR"
+              if $enc_bootloader_elf_not_updated || $enc_bootloader_mbedtls_elf_not_updated; then
+                echo "File $file is in the PR but not enc_bootloader/enc_bootloader.elf"
```
Should this be "...but not enc_bootloader/enc_bootloader.elf or enc_bootloader/enc_bootloader_mbedtls.elf" ?

lurch commented on .github/workflows/check_precompiled.yml:
```diff
@@ -23,50 +26,65 @@ jobs:
-            ALL_FINE=0
-          fi
+          # enc_bootloader
+          enc_bootloader_elf_not_updated=$(! echo "$updated_files" | grep -q "enc_bootloader/enc_bootloader.elf")
```
Might be worth tightening up your grep regexes slightly?
```console
$ echo "FOOenc_bootloader/enc_bootloaderXelfBAR" | grep "enc_bootloader/enc_bootloader.elf"
FOOenc_bootloader/enc_bootloaderXelfBAR
```
:wink:

will-v-pi replied:
I've anchored the starts & ends

lurch commented on .github/workflows/check_precompiled.yml:
```diff
@@ -0,0 +1,97 @@
+                exit 1
+              fi
+            else
+              echo "File $file is not in the PR"
```
Not sure there's much use in printing out filenames which aren't in the PR?

will-v-pi replied:
Sorry, this was just a test commit, hence all the printing - I'll let you know when it's working again for review

lurch commented on .github/workflows/check_precompiled.yml:
```diff
@@ -0,0 +1,97 @@
+              if $flash_id_bin_not_updated; then
+                echo "File $file is in the PR but not picoboot_flash_id/flash_id.bin"
+                exit 1
+              fi
```
~~If I'm understanding the code correctly, then it looks like you could skip creating the `$flash_id_bin_not_updated` variable, and just change the "if" line to~~
```
if $file != "picoboot_flash_id/flash_id.bin"
```
?
~~(IMHO more concise code is generally more readable)~~

EDIT: Oh, now I realise that I was misled by your error message :facepalm: I guess you actually meant something like:
```
echo "File $file is modified in the PR but picoboot_flash_id/flash_id.bin hasn't been updated"
```
?

will-v-pi committed:
Run checks in separate script

will-v-pi committed:
Anchor start/end of greps

lurch commented on .github/workflows/check_precompiled.sh:
```diff
@@ -0,0 +1,51 @@
+# Pass updated files in $updated_files environment variable
+
+# enc_bootloader
+echo "$updated_files" | grep -q "^enc_bootloader/.*\.elf$"
+enc_bootloader_elf_not_updated=$?
```
Should this check that _both_ of the enc_bootloader ELFs have been updated? As it is, I think this check would pass if only one of the ELFs was updated but the other wasn't?

will-v-pi replied:
That was intentional, because you might only change files that are used by one of the versions - this action is mostly a quick double check that you tried to upload changed ELF files, so if you're uploading one and both have changed, you'll hopefully remember to upload the other.

lurch commented on .github/workflows/check_precompiled.sh:
```diff
@@ -0,0 +1,51 @@
+if [ $enc_bootloader_elf_not_updated -eq 1 ]; then
+    echo "Checking enc_bootloader files for modifications as ELFs have not been updated"
+    for file in enc_bootloader/*; do
+        if echo "$file" | grep -q "CMakeLists.txt" || echo "$file" | grep -q "BUILD.bazel"; then
```
I _think_ that this could be simplified to
```shell
if [[ "$file" == "enc_bootloader/CMakeLists.txt" ]] || [[ "$file" == "enc_bootloader/BUILD.bazel" ]]; then
```
?

lurch commented on enc_bootloader/enc_bootloader_mbedtls.elf:
Is the file-permissions change here intentional?

will-v-pi replied:
Thanks for spotting that - it's because of the install step used, I will change that so it installs them all as files rather than programs. Although I'm not sure why all of them aren't showing up as changed modes, as they were all changed to executable like this?

lurch replied:
> Although I'm not sure why all of them aren't showing up as changed modes, as they were all changed to executable like this?

Looks like some of them were already 755, and they've now **all** been set back to 644 :slightly_smiling_face:

will-v-pi committed:
Review fixups

Simplify CMakeLists.txt and BUILD.bazel skips

Install all ELFs as files, so they aren't marked as executable

lurch commented on .github/workflows/check_precompiled.sh:
```diff
@@ -0,0 +1,51 @@
+#!/bin/bash
+
+# Pass updated files in $updated_files environment variable
```
_Conventionally_ environment variables are always created as `$UPPERCASE_VARIABLES`, but for such a simple one-off CI script I guess this doesn't really matter.

will-v-pi replied:
I wanted to treat this more as a local variable anyway - it's only an environment variable to avoid any argument passing issues

lurch commented on .github/workflows/check_precompiled.sh:
```diff
@@ -0,0 +1,51 @@
+        if [[ "$file" == "enc_bootloader/CMakeLists.txt" || "$file" == "enc_bootloader/BUILD.bazel" ]]; then
+            continue
+        fi
+        if echo "$updated_files" | grep -q "$file"; then
```
If you wanted to avoid the use of all these `grep -q` calls, I guess you could use the [builtin bash functionality](https://stackoverflow.com/a/229606) instead (and there's also a [regex-version](https://stackoverflow.com/a/231298) that you could use on e.g. line 6 above).

will-v-pi replied:
That builtin regex functionality doesn't seem to work for the multiline `$updated_files` string, which has one filename per line
(eg `[[ "$updated_files" =~ "^enc_bootloader/.*\.elf$" ]]` always returns `false`)

lurch replied:
Hmmm, perhaps the `^` and `$` are matching start-of-string and end-of-string, rather than start-of-line and end-of-line?
Doesn't really matter though, keep using your existing solution :slightly_smiling_face:

lurch commented:
Hmmm, something that's just occurred to me: if a change was made to a comment in one of the `.c` files, I guess that might result in no change to the generated ELF file; and the CI checks for that comment-fix PR might fail? :thinking:  (but I guess in such a scenario we could just ignore the failing CI and merge anyway?)

will-v-pi commented:
> Hmmm, something that's just occurred to me: if a change was made to a comment in one of the `.c` files, I guess that might result in no change to the generated ELF file; and the CI checks for that comment-fix PR might fail? 🤔 (but I guess in such a scenario we could just ignore the failing CI and merge anyway?)

Yeah, that's something to bear in mind, but I think that should be fine

will-v-pi committed:
Tidy up file checks

lurch commented on .github/workflows/check_precompiled.sh:
```diff
@@ -3,47 +3,41 @@
         fi
-        if echo "$updated_files" | grep -q "$file"; then
+        if [[ "$updated_files" == *"$file"* ]]; then
             echo "File $file is in the PR but flash_id BINs have not been updated"
```
Minor nitpick - is there more than one flash_id BIN?

will-v-pi replied:
Not now, but there might be in the future

lurch commented on .github/workflows/check_precompiled.sh:
```diff
@@ -3,47 +3,41 @@
         fi
-        if echo "$updated_files" | grep -q "$file"; then
+        if [[ "$updated_files" == *"$file"* ]]; then
             echo "File $file is in the PR but xip_ram_perms ELFs have not been updated"
```
Minor nitpick - is there more than one xip_ram_perms ELF?

will-v-pi replied:
Not now, but there might be in the future

kilograham commented on CMakeLists.txt:
```diff
@@ -62,6 +62,12 @@ if (NOT DEFINED USE_PRECOMPILED)
+if (USE_PRECOMPILED)
+    set(EXTERNAL_PROJECT_INSTALL_COMMAND "")
+else()
+    set(EXTERNAL_PROJECT_INSTALL_COMMAND ${CMAKE_COMMAND} --install .)
```
"," gives me the willies

will-v-pi replied:
I've removed that and added a better fix, with a new CMake function to handle the install vs don't install cases

will-v-pi committed:
Add add_embedded_data_project function

This fixes the BUILD_ALWAYS issues as it's now only set when `USE_PRECOMPILED=false`, and also removes the need for `${CMAKE_COMMAND} --install .`

will-v-pi committed:
Add back BUILD_ALWAYS

It's needed so if you update the binaries (eg `git pull`) they get re-copied and re-installed
